components/signoffs/signoffController.js

```diff
@@ -0,0 +1,156 @@
+'use strict';
+
+const signoffDao = require('./signoffDao');
+const JSONStream = require('JSONStream');
+const {
+  handleErrors,
+  ValidationError,
+} = require('../../utils/controllerUtils');
+const { signoff } = require('../../utils/urlSlugs');
+const logger = require('../../utils/logger').logger;
+
+const getSignoff = async (req, res, next) => {
+  try {
+    const signoffs = await signoffDao.getByCaseIdentifier(
+      req.params.caseIdentifier
+    );
+    if (signoffs && signoffs.length) {
+      res.status(200).json(signoffs);
+      next();
+    } else {
+      res.status(404).end();
+    }
+  } catch (e) {
+    handleErrors(e, 'Error getting signoff(s) for case', logger, next);
+  }
+};
+
+const addSignoff = async (req, res, next) => {
+  try {
+    const existingSignoffs = await signoffDao.getByCaseIdentifier(
+      req.body.caseIdentifier
+    );
+
+    const validationResults = validateObjectsFromUser(req.body);
+    if (existingSignoffs == null || !existingSignoffs.length) {
+      const createdSignoff = await upsert(validationResults);
+      return res.status(201).json(createdSignoff);
+    } else {
+      for (const [, existingSignoff] of Object.entries(existingSignoffs)) {
+        if (
+          existingSignoff.signoffStepName ==
+            validationResults.signoffStepName &&
+          existingSignoff.deliverableType == validationResults.deliverableType
+        ) {
+          // signoff step is same, replace the old signoff record
+          const createdSignoff = await upsert(validationResults);
+          return res.status(200).json(createdSignoff);
+        } else {
+          // signoff data is different
+          const createdSignoff = await upsert(validationResults);
+          return res.status(201).json(createdSignoff);
+        }
+      }
```

Comment:
* This is always returning after checking the first record. Any other records are not being checked
* It would be more efficient to create a DAO function that queries by all of these attributes instead of fetching more from the DB and checking them here